torch/monitor: TensorboardEventHandler

d4l3k opened the pull request:
Summary: This adds the beginnings of a TensorboardEventHandler which will log stats to Tensorboard.

Test Plan: buck test //caffe2/test:monitor

Differential Revision: D33719954

edward-io commented on test/test_monitor.py:
```diff
@@ -98,6 +100,60 @@ def handler(event: Event) -> None:
         log_event(e)
         self.assertEqual(len(events), 2)
 
+TEST_TENSORBOARD = True
+try:
+    from torch.utils.tensorboard import SummaryWriter
+    from tensorboard.backend.event_processing import (
+        plugin_event_multiplexer as event_multiplexer,
+    )
+except ImportError:
+    TEST_TENSORBOARD = False
```
can we put this in setUp?

d4l3k replied:
this was copied from the tensorboard tests, but done

edward-io commented on test/test_monitor.py:
```diff
@@ -98,6 +100,60 @@ def handler(event: Event) -> None:
+            return self.skipTest("Skip the test since TensorBoard is not installed")
+        self.temp_dirs = []
+
+    def createSummaryWriter(self):
```
nit: should we stick to snake case for our own methods?

edward-io commented on test/test_monitor.py:
```diff
@@ -98,6 +100,60 @@ def handler(event: Event) -> None:
+        return SummaryWriter(temp_dir.name)
+
+    def tearDown(self):
+        super().tearDown()
```
nit: the parent class doesn't do anything for teardown

edward-io commented on test/test_monitor.py:
```diff
@@ -98,6 +100,60 @@ def handler(event: Event) -> None:
+            temp_dir.cleanup()
+
+    def test_event_handler(self):
+        with self.createSummaryWriter() as w:
```
why do we use a context manager here?

edward-io commented on test/test_monitor.py:
```diff
@@ -98,6 +100,60 @@ def handler(event: Event) -> None:
+            unregister_event_handler(handle)
+
+        mul = event_multiplexer.EventMultiplexer()
+        mul.AddRunsFromDirectory(self.temp_dirs[-1].name, None)
```
the second argument is None by default

edward-io commented on test/test_monitor.py:
```diff
@@ -98,6 +100,60 @@ def handler(event: Event) -> None:
+        }
+        self.assertCountEqual(raw_result.keys(), {"asdf.sum", "asdf.count"})
+        for v in raw_result.values():
+            self.assertEqual(len(v), 5)
```
should we test the actual value here too?

edward-io commented on torch/monitor/__init__.py:
```diff
@@ -1 +1,32 @@
+
+class TensorboardEventHandler:
+    """
+    TensorboardEventHandler is an event handler that will write known events to
+    the provided SummaryWriter.
```
we should also mention it only handles scalars

edward-io replied:
did we update this? :)

d4l3k replied:
see the next line :)

edward-io reviewed (approved):
LGTM, some nits

edward-io reviewed (approved):
lgtm, tiny nit

d4l3k committed:
torch/monitor: TensorboardEventHandler (#71658)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/71658

This adds the beginnings of a TensorboardEventHandler which will log stats to Tensorboard.

Test Plan: buck test //caffe2/test:monitor

Reviewed By: edward-io

Differential Revision: D33719954

fbshipit-source-id: 30c4c1f169f5b42a72684b12e324d418ded656d3

d4l3k commented:
mypy failing due to trunk breakage https://www.torch-ci.com/

https://github.com/pytorch/pytorch/commit/bbe6144b453c7d990bf8c9fb184d14bad30a457d
